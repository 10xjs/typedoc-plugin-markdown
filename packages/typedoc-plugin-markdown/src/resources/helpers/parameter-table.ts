import { ParameterReflection } from 'typedoc';
import { comment } from './comment';
import { stripLineBreaks } from './strip-line-breaks';
import { type } from './type';

export function parameterTable(this: ParameterReflection[]) {
  const defaultValues = this.map((param) => !!param.defaultValue);
  const hasDefaultValues = !defaultValues.every((value) => !value);

  const comments = this.map(
    (param) =>
      (param.comment && !!param.comment.text) ||
      (param.comment && !!param.comment.shortText),
  );
  const hasComments = !comments.every((value) => !value);

  const headers = ['Name', 'Type'];

  if (hasDefaultValues) {
    headers.push('Default');
  }

  if (hasComments) {
    headers.push('Description');
  }

  const rows = this.map((parameter) => {
    const isOptional = parameter.flags.includes('Optional');

    const row = [
      `\`${parameter.flags.isRest ? '...' : ''}${parameter.name}${
        isOptional ? '?' : ''
      }\``,
      type.call(parameter.type),
    ];
    if (hasDefaultValues) {
      row.push(parameter.defaultValue ? parameter.defaultValue : '-');
    }
    if (hasComments) {
      if (parameter.comment) {
        row.push(stripLineBreaks(comment.call(parameter.comment)));
      } else {
        row.push('-');
      }
    }
    return `${row.join(' | ')} |\n`;
  });

  const output = `\n${headers.join(' | ')} |\n${headers
    .map(() => '------')
    .join(' | ')} |\n${rows.join('')}`;

  return output;
}
