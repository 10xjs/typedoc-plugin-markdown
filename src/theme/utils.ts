import * as fs from 'fs';
import * as Handlebars from 'handlebars';
import * as path from 'path';
import { DeclarationReflection, Reflection } from 'typedoc/dist/lib/models/reflections/index';

export function compileTemplate(template: string, data: {}) {
  const source = templateToString(template);
  const md: HandlebarsTemplateDelegate = Handlebars.compile(source);
  return new Handlebars.SafeString(md(data));
}

export function safeString(data: string) {
  return new Handlebars.SafeString(data);
}

export function getAnchorRef(ref: string) {
    return ref.replace(/_|\/|\.| /g, '-').replace(/"/g, '').replace(/ /g, '-').toLowerCase();
  }

function templateToString(template: string) {
  return fs.readFileSync(path.join(__dirname, `/${template}`)).toString();
}
