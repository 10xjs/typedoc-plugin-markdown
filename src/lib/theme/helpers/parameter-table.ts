import * as Handlebars from 'handlebars';
import { ParameterReflection } from 'typedoc';

import { MarkdownPlugin } from '../../plugin';
import { DocusaurusTheme } from '../theme.docusaurus';
import { stripLineBreaks } from './strip-line-breaks';
import { type } from './type';

export function parameterTable(this: ParameterReflection[]) {
  const defaultValues = this.map(param => !!param.defaultValue);
  const hasDefaultValues = !defaultValues.every(value => !value);

  const comments = this.map(
    param => (param.comment && !!param.comment.text) || (param.comment && !!param.comment.shortText),
  );
  const hasComments = !comments.every(value => !value);

  const headers = ['Name', 'Type'];

  if (hasDefaultValues) {
    headers.push('Default');
  }

  if (hasComments) {
    headers.push('Description');
  }

  const rows = this.map(parameter => {
    const isOptional = parameter.flags.includes('Optional');
    const typeOut = type.call(parameter.type);

    const row = [
      `\`${parameter.flags.isRest ? '...' : ''}${parameter.name}${isOptional ? '?' : ''}\``,
      typeOut
        ? typeOut.toString().replace(/\|/g, MarkdownPlugin.theme instanceof DocusaurusTheme ? '&#124;' : '\\|')
        : '',
    ];
    if (hasDefaultValues) {
      row.push(parameter.defaultValue ? parameter.defaultValue : '-');
    }
    if (hasComments) {
      const commentsText = [];
      if (parameter.comment && parameter.comment.shortText) {
        commentsText.push(Handlebars.helpers.comment.call(stripLineBreaks.call(parameter.comment.shortText)));
      }
      if (parameter.comment && parameter.comment.text) {
        commentsText.push(Handlebars.helpers.comment.call(stripLineBreaks.call(parameter.comment.text)));
      }
      row.push(commentsText.length > 0 ? commentsText.join(' ') : '-');
    }
    return `${row.join(' | ')} |\n`;
  });

  const output = `\n${headers.join(' | ')} |\n${headers.map(() => '------').join(' | ')} |\n${rows.join('')}`;

  return output;
}
