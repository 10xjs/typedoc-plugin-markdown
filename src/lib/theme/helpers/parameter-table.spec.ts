import { Fixture, getFixture, getReflectionByName } from '../../../test/test.utils';
import { parameterTable } from './parameter-table';

describe(`parameterTable helper`, () => {
  test(`should compile`, () => {
    const data = getReflectionByName(getFixture(Fixture.Function), 'functionWithDefaults');
    const result = parameterTable.call(data.signatures[0].parameters);
    expect(result).toMatchSnapshot();
  });
});
