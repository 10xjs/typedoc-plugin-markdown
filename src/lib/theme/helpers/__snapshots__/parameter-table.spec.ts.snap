// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parameterTable helper should compile 1`] = `
"
Name | Type | Default | Description |
------ | ------ | ------ | ------ |
\`valueA\` | \`[object Object]\` | \\"defaultValue\\" | A parameter with a default string value. |
\`valueB\` | \`[object Object]\` | 100 | A parameter with a default numeric value. |
\`valueC\` | \`[object Object]\` |  Number.NaN | A parameter with a default NaN value. |
\`valueD\` | \`[object Object]\` | true | A parameter with a default boolean value. |
\`valueE\` | \`[object Object]\` |  null | A parameter with a default null value. |
"
`;
